### DISCUSSION TITLE
find_cells_along_line not working properly

### DISCUSSION THREAD

**Discussion by AsimAl-Sofi** [pyvista/pyvista#4390] [Q&A] [answered]
Hi all
I would like to use the function **find_cells_along_line** to determine the cells along a line, but the function does not give me the right answer.
Here is my code example:
```py
out = pv.CylinderStructured(radius=(0,3),
theta_resolution=10,height=1,
z_resolution=2)
box=out.get_cell(5).get_face(3).cast_to_unstructured_grid().extract_surface()
box.compute_normals(inplace=True)
centers = out.get_cell(5).get_face(3).cast_to_unstructured_grid().cell_centers().points
arrows = box['Normals'][0]
pl = pv.Plotter()
pl.add_mesh(out, opacity=0.6, show_edges=True, line_width=1)
pl.add_points(arrows,color="g", point_size=18.0, render_points_as_spheres=True)
point_S = (0, -5, 2)
pl.add_points(np.array([point_S]),color="r", point_size=18.0, render_points_as_spheres=True)
line=pv.Line(point_S, arrows)
pl.add_mesh(line)
pl.show()
```

the function find_cells_along_line gives me the following answer:
```py
>>> out.find_cells_along_line(point_S, arrows,tolerance=0.0)
array([1, 0, 2, 7, 8, 9])
```
As you can see from the example, the line goes through cell 0,1, 8 and 9 but the function gives 1, 0, 2, 7, 8, 9.
Any suggestions how to tackle this problem?
Thanks

**Comment by banesullivan** [comment]
This method is a thin wrapper around `vtkCellLocator` and indeed this seems like an issue with `vtkCellLocator`

https://github.com/pyvista/pyvista/blob/634c66d9ae7aa81bc413719f2a03a3f6deb9e078/pyvista/core/dataset.py#L2416-L2466

**Reply by MatthewFlamm** [reply]
I've found that if you locate the starting point outside the dataset, it works as expected. This feels like a bug in vtk somewhere as you suggest.

The cells in this case are degenerate Hexahedra, i.e. one quad face is a line.  I tried using `out = pv.CylinderStructured(radius=(0.1,3)`, to make them true Hexahedra, but this did not resolve this issue.

**Reply by AsimAl-Sofi** [reply]
Hi
point_S, i.e., the starting point is already outside the mesh 'out' but still get the wrong answer.
Or do you mean something else?
Thanks

**Reply by MatthewFlamm** [reply]
I'm wondering whether there is no bug, but a mistake in the description of how this method works.  From vtk docs for [FindCellsAlongLine](https://vtk.org/doc/nightly/html/classvtkCellLocator.html#ac6b2fe82d1e42a5c8e7176e561fdb1b5).  My bolding:

- For each intersection with **the bounds** of a cell, the cellIds have the relevant information added sort by t.
  - My reading is that this function only finds cells whose bounds intersect with the line.  It is very probable that when I moved the starting point outside the dataset that I moved it enough to satisfy this condition.

**Reply by MatthewFlamm** [reply]
> Hi
> point_S, i.e., the starting point is already outside the mesh 'out' but still get the wrong answer.
> Or do you mean something else?
> Thanks

I was typing my above comment before you submitted this comment.  I previously meant that both points might need to be outside the dataset.  Starting point was probably  not the best description of this.  But see my new comment, which is probably more relevant.

**Reply by AsimAl-Sofi** [reply]
@MatthewFlamm thank you for the explanation.
That might be indeed the case. However, when looking to the function description, there is a parameter 'tolerance' and when setting to 0.0 I assumed that this will solve the issue but it did not. https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSet.find_cells_along_line.html

**Reply by MatthewFlamm** [reply]
If my hunch is correct, this is the tolerance with respect to the bounds of the cell, i.e. the min/max (x, y, z) positions, not the real shape of the cells themselves.  So setting the tolerance to 0.01 for example would find cells whose bounds are along the line to within +-0.01.  To do what you want, I think we probably need to use `IntersectWithLine` and this is currently not implemented in PyVista.

**Reply by AsimAl-Sofi** [reply]
ah I see
thank you very much for the help

**Reply by MatthewFlamm** [reply, marked as answer]
This behavior can also be observed in our example for `find_cells_along_line`.  A line that travels from inside a sphere to outside should only truly intersect 1 cell.  However, this function returns many cells.  Thanks for raising this issue.  Now tracked #4398

**Reply by AsimAl-Sofi** [reply]
Hi @MatthewFlamm 
I have updated the dataset.py file and looks that the function **find_cells_intersecting_line** works correct.
Thank you for the help
